### ISSUE TITLE
[lambda] Unable to package function requiring compiled NodeJS module

### ISSUE THREAD

**Issue by adam-nielsen** [aws/aws-cdk#11386]
I am trying to deploy a Lambda function that requires the netcdf4 module.  This module is a NodeJS interface to a C library, so installing the module requires some compilation.  I am able to get it to compile but not package up for deployment.  I have a support job open (Case 7555628061) but I thought it would be easier to post the reproduction steps here.

### Reproduction Steps

```
cdk init app --language javascript
npm install @aws-cdk/aws-lambda @aws-cdk/aws-lambda-nodejs
mkdir lambda
cd lambda
npm init  # Accept defaults
echo -e "const netcdf4 = require('netcdf4');\n\nmodule.exports = { handler: () => {} };" > index.js
# You may need to install your distribution's netcdf library before the Node module will compile.
npm install netcdf4
# Previous line fails with recent NodeJS versions, if it fails with compile errors, use git version instead:
# npm install parro-it/netcdf4
cd ..
```
Add this code to lib/cdk-stack.js:

```
const lambdaExample = new NodejsFunction(this, 'example', {
	entry: 'lambda/index.js',
	handler: 'handler',
	runtime: cdkLambda.Runtime.NODEJS_12_X,
	//bundlingDockerImage: cdk.BundlingDockerImage.fromAsset(
	//	'docker/'
	//),
});
```
### What did you expect to happen?

The Lambda function would be packaged up and deployed.

### What actually happened?

Without Docker, I get this error:

    Error: No transformers found for "/asset-input/lambda/node_modules/netcdf4/build/Release/netcdf4.node".

With Docker, I get this error:

```
Bundling asset CdkStack/example/Code/Stage...
internal/modules/cjs/loader.js:638
    throw err;
    ^
Error: Cannot find module 'parcel'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.resolve (internal/modules/cjs/helpers.js:33:19)
    at [eval]:1:9
```

### Environment

  - **CDK CLI Version  :** 1.72.0
  - **Framework Version:**
  - **Node.js Version:** 14.13.0
  - **OS               :** Arch Linux
  - **Language (Version):** JavaScript

### Other

I have attached a [Dockerfile and install scripts](https://github.com/aws/aws-cdk/files/5514053/dockerfile.zip) that can be used to install the netcdf4 dependencies inside a Docker container, so that it is possible to compile the Node module there.  This is probably the preferable solution as it will allow the compiled module to match the Lambda execution environment.

These scripts and Docker container work when using the Serverless Framework application to build and deploy, but I am trying to migrate this to AWS CDK.

--- 

This is :bug: Bug Report

**Comment by jogold** [comment]
You need to use the `nodeModules` option for this. It's a list of modules that should not be bundled but installed instead.

**Comment by adam-nielsen** [comment]
Thanks for getting back to me!

I have added `netcdf4` to the `nodeModules` parameter and now it does treat it differently (attempting to install/compile it rather than bundle it) however I still get an error running `cdk synth` on the host machine:

```
> netcdf4@0.4.1 install /asset-output/node_modules/netcdf4
> node-gyp rebuild

make: Entering directory `/asset-output/node_modules/netcdf4/build'
  CXX(target) Release/obj.target/netcdf4/src/Group.o
../src/Group.cpp:2:10: fatal error: netcdf.h: No such file or directory
 #include <netcdf.h>
```
The file it is looking for exists:
```
$ ls /usr/include/netcdf.h 
-rw-r--r-- 1 root root 67K Apr 11  2020 /usr/include/netcdf.h
```
And I have confirmed it works because the code runs on the local machine and makes use of the module just fine.  However this part I am not so worried about as I believe it is "safer" to have the module built inside Docker so I can use the same glibc version as Lambda to ensure compatibility.  So with that I have tried specifying the same `bundlingDockerImage` as above, but I still get the same error.  In my `Dockerfile` I have:

    RUN npm install -g parcel

Then when I run `cdk synth` I get:

```
Bundling asset CdkStack/example/Code/Stage...
internal/modules/cjs/loader.js:638
    throw err;
    ^
Error: Cannot find module 'parcel'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.resolve (internal/modules/cjs/helpers.js:33:19)
    at [eval]:1:9
[...]
```

I am not sure why it cannot find the `parcel` module when it is available in the container, and indeed if I launch the container I can run `parcel` from the command line as it is in the system path.

What have I missed?

**Comment by jogold** [comment]
Do you need a special Docker image for this? Maybe to compile C?

If no: do not specify the `bundlingDockerImage` prop so that the [default image coming with `@aws-cdk/aws-lambda-nodejs`](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda-nodejs/parcel/Dockerfile) will be used. It is derived from `amazon/aws-sam-cli-build-image-nodejs12.x` with yarn and Parcel added.

If yes: you need to install Parcel in your image the same way it is installed in `@aws-cdk/aws-lambda-nodesjs`

https://github.com/aws/aws-cdk/blob/bf1c1c197bae895bba4ff5088b03d4e9b7d1692f/packages/%40aws-cdk/aws-lambda-nodejs/parcel/Dockerfile#L9-L12

Also if you want to force installing in Docker even if a local version of Parcel is detected, set the `forceDockerBundling` prop to `true`.

**Comment by adam-nielsen** [comment]
Many thanks for this!  This seems to have fixed the problem.

I had to take that Dockerfile, copy and paste it into my own one, and then add my extra commands onto the end to compile and install the libraries.  Once I did this the whole process worked, and I have now successfully deployed the function to Lambda.

I wasn't able to run `yum` for some reason so I couldn't install any packages, but luckily all the packages I already needed were in the base image so everything worked without having to install anything extra.

Hopefully the documentation can be updated to mention the need to copy and paste from this existing Dockerfile if you want to use a custom Docker image.

Thanks again for your help!
